src/runtime_src/core/common/api/xrt_kernel.cpp

```diff
@@ -2056,7 +2056,9 @@ class run_impl
       auto dpu = reinterpret_cast<ert_dpu_data*>(payload);
       dpu->instruction_buffer = addr;
       dpu->instruction_buffer_size = size;
-      dpu->chained = --ert_dpu_data_count;
+      dpu->chained = xrt_core::module_int::get_os_abi(m_module) == xrt_core::module_int::elf_amd_aie2p ?
```

Comment:
This is the main change for this PR

Reply:
I would prefer to have this check within xrt_module implementation.   Can we avoid exposing the constants in module_int.h and instead add an API that fills up the entire payload or minimum returns a bool for either of the decisions taken here.

But based on what I see here, it may actually make more sense to encapsulate everything related to `addr_and_size` and `ert_dpu_data` within xrt_module.cpp.   Simply pass the full payload into an new internal module API, let the implementation collect and walk the `addr_and_size` and fill up the payload.